test/consensus.test.js

```diff
@@ -1723,6 +1724,307 @@ contract('Consensus', async (accounts) => {
+        await mockEoC()
+        await blockReward.cycleMock({from: owner}).should.be.fulfilled
+        if(i < 5)
+        {
+          toBN(i).should.be.bignumber.equal(await consensus.getStrikes(firstCandidate))
+        }
+        else
+        {
+          toBN(5).should.be.bignumber.equal(await consensus.getStrikes(firstCandidate))
+        }
+      }
+
+      currentBlockNumber = await web3.eth.getBlockNumber()
+      currentCycleEndBlock = await consensus.getReleaseBlock(firstCandidate)
+
+      blocksToAdvance = currentCycleEndBlock.toNumber() - currentBlockNumber + 1
+      if(blocksToAdvance > 0) {
+        await advanceBlocks(blocksToAdvance)
+      }
+      await consensus.setBlockCounterMock(secondCandidate,CYCLE_DURATION_BLOCKS);
+      await mockEoC()
+      await blockReward.cycleMock({from: owner}).should.be.fulfilled
+      
+
+      true.should.be.equal(await consensus.isJailed(firstCandidate))
+      false.should.be.equal(await consensus.isPendingValidator(firstCandidate))
+
+      currentBlockNumber = await web3.eth.getBlockNumber()
+      currentCycleEndBlock = await consensus.getReleaseBlock(firstCandidate)
+
+      await consensus.unJail({from: firstCandidate}).should.be.fulfilled
+      await consensus.setBlockCounterMock(firstCandidate,CYCLE_DURATION_BLOCKS);
+      await consensus.setBlockCounterMock(secondCandidate,CYCLE_DURATION_BLOCKS);
+      await mockEoC()
+      await blockReward.cycleMock({from: owner}).should.be.fulfilled
+
+      //check strikes get reset after 50 clean cycles
+      for (i = 0; i <= 50; i++)
+      {
+        await consensus.setBlockCounterMock(firstCandidate,CYCLE_DURATION_BLOCKS);
+        await consensus.setBlockCounterMock(secondCandidate,CYCLE_DURATION_BLOCKS);
+        let currentBlockNumber = await web3.eth.getBlockNumber()
+        let currentCycleEndBlock = await consensus.getCurrentCycleEndBlock()
+        let blocksToAdvance = currentCycleEndBlock.toNumber() - currentBlockNumber
+        await advanceBlocks(blocksToAdvance)
+        true.should.be.equal(await consensus.hasCycleEnded())
+        await blockReward.cycleMock({from: owner}).should.be.fulfilled
+        false.should.be.equal(await consensus.isJailed(firstCandidate))
+        true.should.be.equal(await consensus.isPendingValidator(firstCandidate))
+      }
+      ZERO.should.be.bignumber.equal(await consensus.getStrikes(firstCandidate))
+    })
+  })
+
+  describe('Maintenance', async () => {
+    beforeEach(async () => {
+      await consensus.initialize(initialValidator)
+      await consensus.setProxyStorage(proxyStorage.address)
+    })
+    it('Only validators can flag maintenance', async () => {
+      //test with non val
+      await consensus.maintenance({from: firstCandidate}).should.be.rejectedWith(ERROR_MSG)
+
+      //test with val
+      await consensus.sendTransaction({from: firstCandidate, value: MIN_STAKE}).should.be.fulfilled
```

Comment:
Can you add a test in which validator leaves maintenance mode ?

Reply:
This was already being covered by one of the test, but added some additional checks here e7e799a

Reply:
👍